### PR TITLE
[core] Flush by payload size or time elapsed

### PR DESCRIPTION
**Pull request by brettlangdon**
This PR is investigative work I have done to see what a flushing algorithm based on payload size would look like.

The way we make this work is.

1. When we complete a trace we push it into a queue into a background worker (this is not changing at all).
2. Our background worker continuously pops traces off the queue and adds them to the next payload
  a. When we add to the payload we will encode the trace as JSON or msgpack depending on the encoder being used
3. When the payload size reaches 5mb in size, or after 1 second has passed (which comes first) we will flush the payload to the trace agent
4. Reset the payload and then continue

This PR in it's current state, while the tests are all passing and it seems to work, it feels very messy, there are probably some opportunities to reorganize the code to make it easier to follow.

Highlight of some changes:

- Added in a `Payload` class to keep track of the current payload
  - We have moved filtering of traces into this class as we will apply filters when the trace is added, before it is encoded and stored in the payload
- We have removed/noop'd some "services" stuff
  - This was probably a little unnecessary, easy to add back in
- Updated the `ddtrace.writer.Q`
  - Now inherits from `queue.Queue`
  - Now supports blocking pops

There are some pieces that are still not 100%, like better test cases, doc strings, etc. I wanted to get an overall review before I sank more time into this.

### CODE REVIEW

File: ddtrace/writer.py [hunk 1]
```diff
@@ -112,131 +122,145 @@ def _on_shutdown(self):
...
-                        break
-                if trace is not None:
-                    filtered_traces.append(trace)
-            return filtered_traces
-        return traces
-
-
-class Q(object):
+
+class Q(Queue, object):
```
Comment (Kyle-Verhoog) [inline ddtrace/writer.py]: Definitely need some thorough tests for this bad-boy.

Comment (Kyle-Verhoog) [inline ddtrace/writer.py]: @Kyle-Verhoog note to self: do an in-depth review of the data structure.

File: ddtrace/payload.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    DEFAULT_MAX_PAYLOAD_SIZE = 5 * 1000000
+
+    def __init__(self, encoder=None, filters=None, max_payload_size=DEFAULT_MAX_PAYLOAD_SIZE):
+        self.filters = filters
+        self.max_payload_size = max_payload_size
+        self.encoder = encoder or get_encoder()
+        self.traces = []
+        self.size = 0
+
+    def filter(self, trace):
```
Comment (Kyle-Verhoog) [inline ddtrace/payload.py]: Not sure if I like `filter`s being in the `Payload`. It's clean as is, but doesn't really feel like it's something the payload should be concerned about. To me it seems like a better fit in the writer. But again, since it's clean right now I'm fine either way.

File: ddtrace/writer.py [hunk 3]
```diff
@@ -112,131 +122,145 @@ def _on_shutdown(self):
...
+                    log.debug('Attempting to flush %r', payload)
+                    self._flush(payload)
+                    payload.reset()
+                next_flush = now + 1
+            else:
+                remaining = max(0, next_flush - now)
+                self._trace_queue.wait(remaining)
+
+            # Wait up until the next flush for another trace to be added
+            self._trace_queue.wait(max(0, next_flush - time.time()))
```
Comment (Kyle-Verhoog) [inline ddtrace/writer.py]: :+1:

File: ddtrace/payload.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+class Payload:
+    __slots__ = ['traces', 'size', 'encoder', 'max_payload_size', 'filters']
+
+    # Default max payload size of 5mb
+    # DEV: Trace agent limit is 10mb, cutoff at 5mb to ensure we don't hit 10mb
+    DEFAULT_MAX_PAYLOAD_SIZE = 5 * 1000000
+
+    def __init__(self, encoder=None, filters=None, max_payload_size=DEFAULT_MAX_PAYLOAD_SIZE):
+        self.filters = filters
```
Comment (Kyle-Verhoog) [inline ddtrace/payload.py]: We could have
```suggestion
        self.filters = filters or []
```

then we wouldn't need to `None` check `filters`

File: ddtrace/encoding.py [hunk 5]
```diff
@@ -44,42 +45,78 @@ def encode_traces(self, traces):
...
         return msgpack.packb(obj, **MSGPACK_PARAMS)
 
+    def decode(self, data):
+        return msgpack.unpackb(data)
+
+    def finish(self, objs):
+        buf = b''.join(objs)
+
+        # Prepand array header to buffer
+        # https://github.com/msgpack/msgpack-python/blob/f46523b1af7ff2d408da8500ea36a4f9f2abe915/msgpack/fallback.py#L948-L955
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/encoding.py]: I actually think keeping this here is better, at least until we vendor `msgpack`, since not all versions of `msgpack` have this helper function.

But also, this is exactly the code that `pack_array_header` calls, and the same code exists in `dd-trace-js` and `dd-trace-go`.

File: ddtrace/encoding.py [hunk 6]
```diff
@@ -44,42 +45,78 @@ def encode_traces(self, traces):
...
         raise NotImplementedError
 
+    def decode(self, data):
+        """
+        Defines the underlying format used during traces or services encoding.
+        This method must be implemented and should only be used by the internal functions.
+        """
+        raise NotImplementedError
+
+    def finish(self, objs):
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/encoding.py]: I agree, changing to `join_encoded()` since it is meant to join encoded objects into an array, want to make it more clear.

File: ddtrace/payload.py (hunk 2 above)
Comment (brettlangdon) [reply ddtrace/payload.py]: I agree, moved filter off of the payload and back into writer, that is where it should have been.

File: ddtrace/payload.py (hunk 4 above)
Comment (brettlangdon) [reply ddtrace/payload.py]: I am going to keep as-is, because I would keep the `if not self.filters:` check anyways to check for an empty list.

**Comment by mrmachine** [comment]
@brettlangdon I'm struggling to send traces for Django management commands, taking ~30s to ~10m to execute. I thought I'd give this branch a shot to see if it solves my problems. Unfortunately it did not, and I noticed a possibly undesirable side-effect.

Here's my log from executing the management command code interactively in a Python shell, with IP addresses and hostnames obfuscated:

```
In [2]: create_emu_data.run_create_emu_data()
2019-04-24 13:59:00,134 DEBUG writer.write:36 Queuing trace with 500 spans, 195 additional messages skipped
2019-04-24 13:59:00,814 ERROR writer._flush:201 cannot send spans to OBFUSCATED_DOMAIN:8126: [Errno 32] Broken pipe
2019-04-24 13:59:00,823 DEBUG writer._target:155 Popped trace with 500 spans, 43 additional messages skipped
2019-04-24 13:59:00,835 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107507b, full=False), 6 additional messages skipped
2019-04-24 13:59:00,838 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107507b, full=False), 6 additional messages skipped
2019-04-24 13:59:00,843 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107508b, 6 additional messages skipped
2019-04-24 13:59:01,992 DEBUG api._put:196 Response 200 OK length:488, 5 additional messages skipped
2019-04-24 13:59:01,997 DEBUG api.send_traces:172 Reported 1 traces in 1.15439s, 5 additional messages skipped
2019-04-24 13:59:04,285 DEBUG span.set_tag:138 error setting numeric metric _dd1.sr.eausr:None, 18 additional messages skipped

In [3]: 2019-04-24 13:59:57,735 DEBUG sampler.__init__:41 initialized RateSampler, sample 100% of traces, 6 additional messages skipped
2019-04-24 14:00:00,454 DEBUG api._put:196 Response 200 OK length:521, 43 additional messages skipped
2019-04-24 14:00:00,454 DEBUG api.send_traces:172 Reported 1 traces in 1.37156s, 43 additional messages skipped
2019-04-24 14:00:00,455 DEBUG writer._target:155 Popped trace with 500 spans, 44 additional messages skipped
2019-04-24 14:00:00,457 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107503b, full=False), 44 additional messages skipped
2019-04-24 14:00:00,457 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107503b, full=False), 44 additional messages skipped
2019-04-24 14:00:00,457 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107504b, 44 additional messages skipped
2019-04-24 14:00:05,100 DEBUG sampler.__init__:41 initialized RateSampler, sample 100% of traces
In [3]:

In [3]: 2019-04-24 14:01:00,554 DEBUG api._put:196 Response 200 OK length:674, 40 additional messages skipped
2019-04-24 14:01:00,554 DEBUG api.send_traces:172 Reported 1 traces in 2.08556s, 40 additional messages skipped
2019-04-24 14:01:00,556 DEBUG writer._target:155 Popped trace with 500 spans, 40 additional messages skipped
2019-04-24 14:01:00,563 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107503b, full=False), 40 additional messages skipped
2019-04-24 14:01:00,564 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107503b, full=False), 40 additional messages skipped
2019-04-24 14:01:00,564 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107504b, 40 additional messages skipped
2019-04-24 14:01:28,364 DEBUG api.get_json:86 Unable to parse Datadog Agent JSON response: Expecting value: line 1 column 1 (char 0) 'read tcp OBFUSCATED_IP:8126->OBFUSCATED_IP:62946: i/o timeout\n'
2019-04-24 14:01:28,365 ERROR writer._log_error_status:218 failed_to_send traces to Datadog Agent: HTTP error status 400, reason Bad Request, message Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Wed, 24 Apr 2019 04:31:28 GMT
Content-Length: 59


In [3]:
Do you really want to exit ([y]/n)?
2019-04-24 14:01:58,199 DEBUG span.set_tag:138 error setting numeric metric _dd1.sr.eausr:None, 103 additional messages skipped
2019-04-24 14:01:58,200 DEBUG writer.write:36 Queuing trace with 1 spans, 194 additional messages skipped
2019-04-24 14:01:58,410 DEBUG writer._on_shutdown:115 Waiting 5s for traces to be sent. Hit ctrl-c to quit.
2019-04-24 14:02:01,175 DEBUG api._put:196 Response 200 OK length:673, 32 additional messages skipped
2019-04-24 14:02:01,176 DEBUG api.send_traces:172 Reported 1 traces in 1.73121s, 32 additional messages skipped
2019-04-24 14:02:01,177 DEBUG writer._target:155 Popped trace with 500 spans, 32 additional messages skipped
2019-04-24 14:02:01,181 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107503b, full=False), 32 additional messages skipped
2019-04-24 14:02:01,181 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107503b, full=False), 32 additional messages skipped
2019-04-24 14:02:01,181 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107504b, 32 additional messages skipped
2019-04-24 14:03:00,046 DEBUG api._put:196 Response 200 OK length:675, 37 additional messages skipped
2019-04-24 14:03:00,047 DEBUG api.send_traces:172 Reported 1 traces in 1.36665s, 37 additional messages skipped
2019-04-24 14:03:00,048 DEBUG writer._target:155 Popped trace with 500 spans, 37 additional messages skipped
2019-04-24 14:03:00,051 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107505b, full=False), 37 additional messages skipped
2019-04-24 14:03:00,052 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107505b, full=False), 37 additional messages skipped
2019-04-24 14:03:00,052 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107506b, 37 additional messages skipped
2019-04-24 14:04:02,631 DEBUG api._put:196 Response 200 OK length:554, 42 additional messages skipped
2019-04-24 14:04:02,632 DEBUG api.send_traces:172 Reported 1 traces in 3.70227s, 42 additional messages skipped
2019-04-24 14:04:02,633 DEBUG sampler.__init__:41 initialized RateSampler, sample 25.329419826708303% of traces, 2 additional messages skipped
2019-04-24 14:04:02,633 DEBUG writer._target:155 Popped trace with 500 spans, 42 additional messages skipped
2019-04-24 14:04:02,634 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107505b, full=False), 42 additional messages skipped
2019-04-24 14:04:02,635 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107505b, full=False), 42 additional messages skipped
2019-04-24 14:04:02,635 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107506b, 42 additional messages skipped
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/Users/OBFUSCATED/src/ddtrace/ddtrace/writer.py", line 118, in _on_shutdown
    self._trace_queue.join()
  File "/Users/OBFUSCATED/src/ddtrace/ddtrace/writer.py", line 334, in join
    self.not_full.wait(0.05)
  File "/Users/OBFUSCATED/.pyenv/versions/3.6.6/lib/python3.6/threading.py", line 299, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt

$ 
```

I was already getting the broken pipe errors before I tried this branch. I had hoped this branch would fix them. And other traces are being sent correctly (`django.request`, `sqlite`, etc.) So I thought it was just related to the size of the trace for this long running code.

But the possibly undesirable thing I noticed with this branch (or at least a different thing vs ddtrace-py 0.24) is that it keeps trying to send the spans again and again after the command has finished executing, and when I attempt to exit the interactive shell it says it will try for 5s, but it actually just keeps trying forever until `KeyboardInterrupt`.

I also set `export DD_TRACER_PARTIAL_FLUSH_ENABLED=true` before opening this Python shell, which didn't seem to make any difference.

**Comment by brettlangdon** [comment]
@mrmachine I opened up #898 so we can triage there instead of in this PR.

File: ddtrace/writer.py [hunk 7]
```diff
@@ -169,64 +218,117 @@ def _log_error_status(self, response, response_name):
...
+            # DEV: `with self.not_empty` will acquire a lock on `self.mutex`
+            with self.not_empty:
+                self.not_empty.wait(seconds)
+
+    def join(self):
+        """
+        Block until all items are removed from the queue
+        """
+        # Wait until after all items have been removed from the queue
+        with self.not_full:
```
Comment (brettlangdon) [inline ddtrace/writer.py]: Update this to also only block for a max of 5 seconds.

File: ddtrace/writer.py [hunk 8]
```diff
@@ -105,51 +106,99 @@ def _on_shutdown(self):
...
-                    time.sleep(0.05)
+                # Block until all items have been removed from the queue
+                self._trace_queue.join()
+
+    def _filter_trace(self, trace):
+        if not trace:
+            return None
+
+        if not self._filters:
+            return trace
```
Comment (jd) [inline ddtrace/writer.py]: This can be safely removed and would simplify reading the code.

File: ddtrace/writer.py [hunk 9]
```diff
@@ -105,51 +106,99 @@ def _on_shutdown(self):
...
+            trace = f.process_trace(trace)
+            if trace is None:
+                log.debug('filter %r filtered out trace %r', f, trace)
+                return None
+
+        return trace
 
     def _target(self):
-        traces_response = None
+        next_flush = time.time() + 1
```
Comment (jd) [inline ddtrace/writer.py]: Maybe put that `1` in a class variable?

File: ddtrace/writer.py [hunk 10]
```diff
@@ -105,51 +106,99 @@ def _on_shutdown(self):
...
+                    log.debug('Attempting to flush %r', payload)
+                    self._flush(payload)
+
+                    # Get a fresh payload
+                    # DEV: Payloads are not reusable
+                    payload = self.api.new_payload()
+                next_flush = now + 1
+
+            # Wait up until the next flush for another trace to be added
+            self._trace_queue.wait(max(0, next_flush - now))
```
Comment (jd) [inline ddtrace/writer.py]: I don't know how precise you want to be, but the flushing could take some time, so your `now` might be a little be outdated at that point.

File: ddtrace/writer.py [hunk 11]
```diff
@@ -105,51 +106,99 @@ def _on_shutdown(self):
...
 
-            if self._priority_sampler and traces_response:
-                result_traces_json = traces_response.get_json()
-                if result_traces_json and 'rate_by_service' in result_traces_json:
-                    self._priority_sampler.set_sample_rate_by_service(result_traces_json['rate_by_service'])
+    def _flush(self, payload):
+        # Nothing to do, return early
+        if payload.empty:
+            log.debug('Payload empty, not flushing %r', payload)
+            return
```
Comment (jd) [inline ddtrace/writer.py]: On L161 and L176 there are already emptiness checks.
I feel like it should be decided where the check should be done and done only once and for all.

File: ddtrace/writer.py [hunk 12]
```diff
@@ -169,64 +218,117 @@ def _log_error_status(self, response, response_name):
...
-            return len(self._things)
+        :param maxsize: The max number of items to put into the queue
+        :type maxsize: int
+        """
+        super(Q, self).__init__(maxsize=maxsize)
+        self._closed = False
 
     def close(self):
-        with self._lock:
+        """
```
Comment (jd) [inline ddtrace/writer.py]: You'll see while rebasing, but the `close` attribute is actually useless the way the Q is used.

File: ddtrace/writer.py [hunk 13]
```diff
@@ -169,64 +218,117 @@ def _log_error_status(self, response, response_name):
...
+    def join(self):
+        """
+        Block until all items are removed from the queue
+        """
+        # Wait until after all items have been removed from the queue
+        with self.not_full:
+            # DEV: Do not use `self.empty()` or `self.qsize()` here since we already have a lock on `self.mutex`
+            while self._qsize():
+                # DEV: `not_full` is notified when we remove items from the queue
+                self.not_full.wait(0.05)
```
Comment (jd) [inline ddtrace/writer.py]: Not a fan of this. Why not reusing `self.all_tasks_done`?

File: ddtrace/encoding.py [hunk 14]
```diff
@@ -9,11 +9,13 @@
 try:
     from ddtrace.vendor import msgpack
     # DEV: `use_bin_type` only exists since `0.4.0`, but we vendor a more recent version
-    MSGPACK_PARAMS = {'use_bin_type': True}
+    MSGPACK_PACK_PARAMS = {'use_bin_type': True}
+    MSGPACK_UNPACK_PARAMS = {'raw': False}
```
Comment (jd) [inline ddtrace/encoding.py]: Why?

File: ddtrace/encoding.py [hunk 15]
```diff
@@ -9,11 +9,13 @@
 try:
     from ddtrace.vendor import msgpack
     # DEV: `use_bin_type` only exists since `0.4.0`, but we vendor a more recent version
-    MSGPACK_PARAMS = {'use_bin_type': True}
```
Comment (jd) [inline ddtrace/encoding.py]: So that was unused right?

I feel like fixing that would deserve a different PR, but ok.

File: ddtrace/api.py [hunk 16]
```diff
@@ -136,7 +136,7 @@ def _set_version(self, version, encoder=None):
         # overwrite the Content-type with the one chosen in the Encoder
         self._headers.update({'Content-Type': self._encoder.content_type})
 
-    def _downgrade(self):
+    def _downgrade(self, payload):
```
Comment (jd) [inline ddtrace/api.py]: Looks wrong.

Comment (brettlangdon) [reply ddtrace/api.py]: yes it does

**Comment by brettlangdon** [comment]
Closing in favor of #941
